File: src/xdevs/core/lib/examples/Transducer.java
package xdevs.core.lib.examples;

import java.util.LinkedList;
import java.util.logging.Logger;

import xdevs.core.modeling.Atomic;
import xdevs.core.modeling.InPort;
import xdevs.core.modeling.OutPort;

/**
 *
 * @author José Luis Risco Martín TODO: I keep the Transducer atomic model for
 * the end ...
 */
public class Transducer extends Atomic {

    private static final Logger logger = Logger.getLogger(Transducer.class.getName());

    protected InPort<Job> iArrived = new InPort<>("iArrived");
    protected InPort<Job> iSolved = new InPort<>("iSolved");
    protected OutPort<Job> oOut = new OutPort<>("oOut");
    protected OutPort<Result> oResult = new OutPort<>("oResult");

    protected LinkedList<Job> jobsArrived = new LinkedList<>();
    protected LinkedList<Job> jobsSolved = new LinkedList<>();
    protected double observationTime;
    protected double totalTa;
    protected double clock;
    protected Result result;

    public Transducer(String name, double observationTime) {
        super(name);
        super.addInPort(iArrived);
        super.addInPort(iSolved);
        super.addOutPort(oOut);
        super.addOutPort(oResult);
        totalTa = 0;
        clock = 0;
        this.observationTime = observationTime;
    }

    @Override
    public void initialize() {
        super.holdIn("active", observationTime);
    }

    @Override
    public void deltint() {
        clock = clock + getSigma();
        double throughput;
        double avgTaTime;
        if (phaseIs("active")) {
            if (!jobsSolved.isEmpty()) {
                avgTaTime = totalTa / jobsSolved.size();
                if (clock > 0.0) {
                    throughput = jobsSolved.size() / clock;
                } else {
                    throughput = 0.0;
                }
            } else {
                avgTaTime = 0.0;
                throughput = 0.0;
            }
            logger.info("End time: " + clock);
            logger.info("Jobs arrived : " + jobsArrived.size());
            logger.info("Jobs solved : " + jobsSolved.size());
            logger.info("Average TA = " + avgTaTime);
            logger.info("Throughput = " + throughput);
            result = new Result(throughput, avgTaTime, jobsArrived.size(), jobsSolved.size());
            holdIn("done", 0);
        } else {
            passivate();
        }
        //logger.info("####deltint: "+showState());
    }

    @Override
    public void deltext(double e) {
        clock = clock + e;
        if (phaseIs("active")) {
            Job job = null;
            if (!iArrived.isEmpty()) {
                job = iArrived.getSingleValue();
                logger.fine("Start job " + job.id + " @ t = " + clock);
                job.time = clock;
                jobsArrived.add(job);
            }
            if (!iSolved.isEmpty()) {
                job = iSolved.getSingleValue();
                totalTa += (clock - job.time);
                logger.fine("Finish job " + job.id + " @ t = " + clock);
                job.time = clock;
                jobsSolved.add(job);
            }
        }
        //logger.info("###Deltext: "+showState());
    }

    @Override
    public void lambda() {
        if (phaseIs("done")) {
            Job job = new Job("null");
            oOut.addValue(job);
            oResult.addValue(result);
            logger.info("" + result);
        }
    }
}

File: examples/efp/transducer.go
/*
 * Copyright (c) 2021, Román Cárdenas Rodríguez.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package efp

import (
	"fmt"
	"github.com/dacya/xdevs/pkg/modeling"
	"github.com/dacya/xdevs/pkg/util"
)

type Transducer struct {
	modeling.Atomic
	iArrived        modeling.Port
	iSolved         modeling.Port
	oOut            modeling.Port
	jobsArrived     []Job
	jobsSolved      []Job
	observationTime float64
	totalTA         float64
	clock           float64
}

func NewTransducer(name string, observationTime float64) *Transducer {
	t := Transducer{
		modeling.NewAtomic(name),
		modeling.NewPort("iArrived", make([]Job, 0)),
		modeling.NewPort("iSolved", make([]Job, 0)),
		modeling.NewPort("oOut", make([]Job, 0)),
		nil,
		nil,
		observationTime,
		0,
		0,
	}
	t.AddInPort(t.iArrived)
	t.AddInPort(t.iSolved)
	t.AddOutPort(t.oOut)
	return &t
}

func (t *Transducer) Initialize() {
	t.HoldIn(util.ACTIVE, t.observationTime)
}

func (t *Transducer) Exit() {}

func (t *Transducer) DeltInt() {
	t.clock += t.GetSigma()
	throughput, avgTaTime := 0.0, 0.0
	if t.GetPhase() == util.ACTIVE {
		if len(t.jobsSolved) > 0 {
			avgTaTime = t.totalTA / float64(len(t.jobsSolved))
			if t.clock > 0 {
				throughput = float64(len(t.jobsSolved)) / t.clock
			}
		}
		fmt.Printf("End time: %v\n", t.clock)
		fmt.Printf("Jobs arrived: %v\n", len(t.jobsArrived))
		fmt.Printf("Jobs solved: %v\n", len(t.jobsSolved))
		fmt.Printf("Average TA: %v\n", avgTaTime)
		fmt.Printf("Throughput: %v\n", throughput)
		t.HoldIn("done", 0)
	} else {
		t.Passivate()
	}
}

func (t *Transducer) DeltExt(e float64) {
	t.clock += e
	if t.PhaseIs(util.ACTIVE) {
		job := Job{}
		if !t.iArrived.IsEmpty() {
			job = t.iArrived.GetSingleValue().(Job)
			job.Time = t.clock
			fmt.Printf("Start job %v @ t = %v\n", job.Id, t.clock)
			t.jobsArrived = append(t.jobsArrived, job)
		}
		if !t.iSolved.IsEmpty() {
			job = t.iSolved.GetSingleValue().(Job)
			t.totalTA += t.clock - job.Time
			fmt.Printf("Finish job %v @ t = %v\n", job.Id, t.clock)
			job.Time = t.clock
			t.jobsSolved = append(t.jobsSolved, job)
		}
	}
	t.HoldIn(t.GetPhase(), t.GetSigma()-e)
}

func (t *Transducer) DeltCon(e float64) {
	t.DeltInt()
	t.DeltExt(0)
}

func (t *Transducer) Lambda() {
	if t.PhaseIs("done") {
		t.oOut.AddValue(Job{})
	}
}
